### PR TITLE
Add docstrings to submesh creation in Python interface

### PR DESCRIPTION
**Pull request by jorgensd**
Currently not exposed in documentation.

Also adds proper wrappers around the maps returned in submesh creation.

### CODE REVIEW

File: python/dolfinx/mesh.py [hunk 1]
```diff
@@ -439,7 +440,20 @@ def create_mesh(
     return Mesh(mesh, domain)
 
 
-def create_submesh(msh, dim, entities):
+def create_submesh(msh: Mesh, dim: int, entities: npt.NDArray[np.int32]):
+    """
+    Create a mesh based on a subset of entities from an existing mesh.
```
Comment (garth-wells) [inline python/dolfinx/mesh.py:445]: This should be one line up. Something must be missing in the ruff checks to not pick this up.

File: python/dolfinx/mesh.py [hunk 2]
```diff
@@ -439,7 +440,20 @@ def create_mesh(
...
+
+    Args:
+        mesh: Input mesh
+        dim: Topological dimension of the entities to extract
+        entities: Indices of entities to extract
```
Comment (garth-wells) [inline python/dolfinx/mesh.py:450]:
```suggestion
        entities: Indices of entities in ``mesh`` to extract
```

File: python/dolfinx/mesh.py [hunk 3]
```diff
@@ -439,7 +440,20 @@ def create_mesh(
...
+    Returns:
+        A quadrulet containing the submesh, a map from the cells of the submesh
```
Comment (garth-wells) [inline python/dolfinx/mesh.py:452]: I think this needs rephrasing and restructuring - it's very hard to follow.

**Review by garth-wells — CHANGES_REQUESTED** [review]

File: python/dolfinx/mesh.py [hunk 4]
```diff
@@ -439,7 +440,18 @@ def create_mesh(
...
+def create_submesh(msh: Mesh, dim: int, entities: npt.NDArray[np.int32]):
+    """Create a mesh based on a subset of entities from an existing mesh.
+
+    Args:
+        mesh: Input mesh
+        dim: Topological dimension of the entities to extract
+        entities: Indices of entities in ``mesh`` to extract
+    Returns:
+        A quadruplet containing the submesh and three maps from the submesh to the parent mesh.
+        The first map is the entity map, the second map the vertex map (topology) and the third map
```
Comment (jorgensd) [inline python/dolfinx/mesh.py:452]:
```suggestion
        The first map is the entity map, the second map the vertex map (topology), and the third map
```

**Comment by jhale** [comment]
Please merge this once you are happy with it.

File: python/dolfinx/mesh.py [hunk 5]
```diff
@@ -439,7 +440,18 @@ def create_mesh(
     return Mesh(mesh, domain)
 
 
-def create_submesh(msh, dim, entities):
+def create_submesh(msh: Mesh, dim: int, entities: npt.NDArray[np.int32]):
+    """Create a mesh based on a subset of entities from an existing mesh.
```
Comment (garth-wells) [inline python/dolfinx/mesh.py:444]:
```suggestion
    """Create a mesh that with specified entities from another mesh.
```

File: python/dolfinx/mesh.py [hunk 6]
```diff
@@ -439,7 +440,18 @@ def create_mesh(
...
+
+    Args:
+        mesh: Input mesh
```
Comment (garth-wells) [inline python/dolfinx/mesh.py:447]:
```suggestion
        mesh: Mesh to create the sub-mesh from.
```

File: python/dolfinx/mesh.py [hunk 7]
```diff
@@ -439,7 +440,18 @@ def create_mesh(
...
+        dim: Topological dimension of the entities to extract
```
Comment (garth-wells) [inline python/dolfinx/mesh.py:448]:
```suggestion
        dim: Topological dimension of the entities in ``msh`` to include in the sub-mesh.
```

File: python/dolfinx/mesh.py [hunk 8]
```diff
@@ -439,7 +440,18 @@ def create_mesh(
...
+        entities: Indices of entities in ``msh`` to extract
```
Comment (garth-wells) [inline python/dolfinx/mesh.py:449]:
```suggestion
        entities: Indices of entities in ``msh`` to include in the sub-mesh.
```

File: python/dolfinx/mesh.py [hunk 9]
```diff
@@ -439,7 +440,18 @@ def create_mesh(
...
+    Returns:
+        A quadruplet containing the submesh and three maps from the submesh to
```
Comment (garth-wells) [inline python/dolfinx/mesh.py:451]: The return type is a `tuple`. Use (0), (1), etc, to be precise on what each element is. 'map' needs to be clear on what it maps from and to.

**Review by garth-wells — CHANGES_REQUESTED** [review]

File: python/dolfinx/mesh.py [hunk 10]
```diff
@@ -439,7 +440,18 @@ def create_mesh(
     return Mesh(mesh, domain)
 
 
-def create_submesh(msh, dim, entities):
+def create_submesh(msh: Mesh, dim: int, entities: npt.NDArray[np.int32]):
```
Comment (garth-wells) [inline python/dolfinx/mesh.py:443]: Add return type hint.

File: python/dolfinx/mesh.py (hunk 5 above)
Comment (jorgensd) [reply python/dolfinx/mesh.py:444]: To me that new sentence doesn’t make sense.

File: python/dolfinx/mesh.py [hunk 11]
```diff
@@ -439,7 +440,19 @@ def create_mesh(
...
+    msh: Mesh, dim: int, entities: npt.NDArray[np.int32]
+) -> tuple[Mesh, npt.NDArray[np.int32], npt.NDArray[np.int32], npt.NDArray[np.int32]]:
+    """Create a mesh with specified entities from another mesh.
+
+    Args:
+        mesh: Mesh to create the sub-mesh from.
+        dim: Topological dimension of the entities in ``msh`` to include in the sub-mesh.
+        entities: Indices of entities in ``msh`` to include in the sub-mesh.
+    Returns:
+        The (1) sub mesh, (2) entity map, (3) vertex map and (4) node map (geometry).
```
Comment (jorgensd) [inline python/dolfinx/mesh.py:453]:
```suggestion
        The (1) sub mesh, (2) entity map, (3) vertex map, and (4) node map (geometry).
```

**Review by garth-wells — APPROVED** [review]
